// @wdio/browserstack-service/build/testOps/featureUsage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../util");
class FeatureUsage {
    constructor(isTriggered) {
        if (isTriggered !== undefined) {
            this.isTriggered = isTriggered;
        }
    }
    getTriggered() {
        return this.isTriggered;
    }
    setTriggered(triggered) {
        this.isTriggered = triggered;
    }
    setStatus(status) {
        this.status = status;
    }
    setError(error) {
        this.error = error;
    }
    triggered() {
        this.isTriggered = true;
    }
    failed(e) {
        this.status = 'failed';
        this.error = (0, util_1.getErrorString)(e);
    }
    success() {
        this.status = 'success';
    }
    getStatus() {
        return this.status;
    }
    getError() {
        return this.error;
    }
    toJSON() {
        return {
            isTriggered: this.isTriggered,
            status: this.status,
            error: this.error
        };
    }
}
exports.default = FeatureUsage;
