// @wdio/browserstack-service/build/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const testOpsConfig_1 = __importDefault(require("./testOps/testOpsConfig"));
const util_1 = require("./util");
class BrowserStackConfig {
    static getInstance(options, config) {
        if (!this._instance && options && config) {
            this._instance = new BrowserStackConfig(options, config);
        }
        return this._instance;
    }
    constructor(options, config) {
        this.funnelDataSent = false;
        this.framework = config.framework;
        this.userName = config.user;
        this.accessKey = config.key;
        this.testObservability = new testOpsConfig_1.default(options.testObservability !== false, !(0, util_1.isUndefined)(options.testObservability));
        this.percy = options.percy || false;
        this.accessibility = options.accessibility || false;
        this.app = options.app;
        this.appAutomate = !(0, util_1.isUndefined)(options.app);
        this.automate = !this.appAutomate;
        this.buildIdentifier = options.buildIdentifier;
    }
    sentFunnelData() {
        this.funnelDataSent = true;
    }
}
exports.default = BrowserStackConfig;
